### PR TITLE
Deduplicate Only During Pre-Processing

### LINKED ISSUE #2719: Save De-duplicated Set During Pre-processing
### ⚠️ Please check that this feature request hasn't been suggested before.

- [x] I searched previous [Ideas in Discussions](https://github.com/axolotl-ai-cloud/axolotl/discussions/categories/ideas) didn't find any similar feature requests.
- [x] I searched previous [Issues](https://github.com/axolotl-ai-cloud/axolotl/labels/enhancement) didn't find any similar feature requests.

### 🔖 Feature description

Currently the de-duplication is being run *after* the set has been saved during the pre-processing phase, and gets run again when starting the training phase (as in doing `-m axolotl.cli.preprocess` then doing `-m axolotl.cli.train`).

For small sets it might not matter much, but for larger sets (which would make more sense to be de-duplicating) this takes a *long* time. That means that if you had a large dataset that took like 15 minutes to de-duplicate you'd be spending 30 minutes on just de-duplicating, before you get to take a single training step, plus another 15 minutes any time you wanted to restart training for whatever reason, like OOM, hyper-parameter tuning, etc.

```
Example time to de-duplicate one of my configs: ~14 minutes.

Saving the dataset (23/23 shards): 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████| 162329/162329 [01:35<00:00, 1694.97 examples/s]
[2025-05-23 22:41:55,123] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:134] [PID:122699] [RANK:0] Train dataset is None. Skipping deduplication.
[2025-05-23 22:41:55,123] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:147] [PID:122699] [RANK:0] Eval dataset is None. Skipping deduplication.
[2025-05-23 22:41:55,124] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:150] [PID:122699] [RANK:0] Starting deduplication for combined dataset. Original size: 162329
[2025-05-23 22:56:28,138] [INFO] [axolotl.utils.data.utils.deduplicate_and_log_datasets:154] [PID:122699] [RANK:0] Deduplication complete for combined dataset. New size: 159750
```

If this de-duplication step was instead done *only* during the pre-processing stage and then *saved*, you would only need to do it once and would also be able to push the de-duplicated set to huggingface.

### ✔️ Solution

Move the de-duplication to *before* saving in the pre-processing phase. Also prevent it from being run before the training phase.

### ❓ Alternatives

Not exactly an alternative, but an addition; making the de-duplication code multi-threaded. Currently it only seems to hit 1 or 2 threads. So that's probably where a lot of the slowness is coming from.

### 📝 Additional Context

_No response_

### Acknowledgements

- [x] My issue title is concise, descriptive, and in title casing.
- [x] I have searched the existing issues to make sure this feature has not been requested yet.
- [x] I have provided enough information for the maintainers to understand and evaluate this request.

### PR DESCRIPTION
**Pull request by xzuyn**
Resolves #2719

## Summary by CodeRabbit

- **Refactor**
  - Improved dataset deduplication by centralizing the deduplication process, ensuring it occurs only once after datasets are merged and optionally shuffled.
  - Simplified internal workflow for handling datasets, reducing redundant operations.

### CODE REVIEW

**Comment by xzuyn** [comment]
Closing in favour of #2747
